### DISCUSSION TITLE
why daisyui add base classes not utility classes to the component?

### DISCUSSION THREAD

**Discussion by yaman3bd** [saadeghi/daisyui#1782] [General]
I was wondering why would `daisyui` use a base parent class like `btn btn-{color}` etc...
not used utility classes directly to the element, even Tailwind says not to use such base classes because the compiled .css file would increase in size.

for example, the `btn` class has `display: flex` and I want to add `flex` class to another element the `display: flex` property will be added to `btn` and `flex` classes instead of only one class and that will result in the compiled .css file to increase in size.
also, I asked in the Tailwind discord community everyone said they would use utility classes.

I am working on React component library for internal use, but I am really confused about what to use base parent classes or utility classes, and I spent a long time searching on what is the best approach to follow but I came across the issue of increasing in size of the compiled .css file or conflict with tailwind classes
so I am really curious to understand how `daisyui` solved this issue.

**Comment by saadeghi** [comment]
Apologies for long answer but you asked a general question and there's a lot to talk about here so I'm going to slice it to smaller questions:

#### Utility, Component... What's the difference?
There is no "best approach". It's just a different level of abstraction.  
- If you write inline CSS, it's the lowest level of abstraction
- If you write only utility classes, it's a little bit higher
- If you write component classes like Bootstrap, it's a higher level of abstraction
- If you import a big JS component which has a lot of HTML elements and styles included (for example `<DatasetTable tabledata={obj}>`) it's the highest level

As you can see:
- Lower levels give you more customizability and you can control everything but you have to write more and lowers your development speed.
- Higher levels give you faster development speed but you have less control on little details

#### So how is daisyUI helping?

daisyUI is not a replacement for Tailwind CSS. It's an addition because with Tailwind CSS + daisyUI you can create pages faster and you can still customize the details using utility classes.

#### But how about smaller CSS size?

I can confirm using daisyUI increases the CSS file size (about 5% to 15% depending on the project) but of course it lowers the markup size about the same amount.  
Let's say we used 100 utility classes total with utility-only approach, our CSS will have 100 class names, each one for a CSS rule. However, we use these class names in the HTML multiple times. So with using Tailwind CSS (no daisyUI) we lowered the size of CSS in cost of increasing the size of HTML.

#### Speed of development?

Tailwind CSS is suggesting a utility-first approach to maximize the customizability but that's not the fastest way to develop apps. You can try using Tailwind CSS (with no component library) to create a few websites, after a while you will ask yourself "why should I create the same `toggle` style for all my projects again and again?" because there is a high chance all your projects need a standard set of common styles for most of the elements anyway. You need the styles for button, checkbox, menu, etc. again and again. The only difference is probably colors or border-radius. So it's not efficient to copy/paste 60 utility classes through all your projects again and again to re-create the same button style.  

In my experience if you want to develop faster, you would need a higher level of abstraction so instead of writing 60 utility classes for a button, you can just quickly write `btn` and you get pre-designed button with hover effect, focus effect, sizes and colors. Then you can add a `btn-primary` to modify the role of your button or `btn-lg` to modify the size (these are the standard design tokens you will need in your design system anyway) And thanks to Tailwind CSS' utility classes you can customize them even more if you need.  

#### But is daisyUI against the utility-first approach?

And I should mention that daisyUI or any component library on top of Tailwind CSS is not **"against the utility-first approach"** A daisyUI project is still utility-first but there are additional components for faster development because we don't want to re-invent the wheel for all the common elements.  
Having components in Tailwind CSS is a feature from Tailwind CSS itself. As you can see [here](https://tailwindcss.com/docs/plugins) there is API for adding component classes and you can even see a `.btn` example. Also there are "component library" plugins created and recommended by Tailwind CSS itself like [`tailwindcss-forms`](https://github.com/tailwindlabs/tailwindcss-forms)

#### But we already experienced Bootstrap... are we going full circle?   

Bootstrap's idea of having class names for elements was not bad at all. The problem with bootstrap was lack of customizability. That's why all Bootstrap websites look like each other and the only way to build something different was to write lots of custom CSS to override the default theme.  
It's not the same with daisyUI because daisyUI itself is on top of Tailwind CSS so we have the power to customize everything using CSS classes. Also, with daisyUI's theme engine you can be sure your project has a unique look comparing to all other websites built with daisyUI.

#### But is it a big commitment to use daisyUI instead of utility classes?

It's not a big commitment. You don't have to use daisyUI for everything. For example you can use 2 or 3 components and keep using Tailwind CSS utility classes for everything else. Unused styles will be purged by Tailwind anyway.

#### I'm not sure yet...

If you're not sure if you need daisyUI or not, I would suggest keep using utility classes only for 5~6 projects. That would be the point you will come to the conclusion that you need a faster way.

**Reply by ltbits** [reply]
> And I should mention that daisyUI or any component library on top of Tailwind CSS is not "against the utility-first approach"

Unfortunately it is. It brings back that "Content-agnostic CSS components" class naming problem that [Tailwind was originally created](https://adamwathan.me/css-utility-classes-and-separation-of-concerns/) to solve.

> Having components in Tailwind CSS is a feature from Tailwind CSS itself.

This argument falls short. Tailwind has its official (paid) [component library](https://tailwindcss.com/plus/ui-blocks/application-ui) and they're built entirely with utility class names, there's no '.button', '.modal' and etc.

**Reply by saadeghi** [reply]
@ltbits If using utility classes only is working for you, keep using them. I'm not going to convince you otherwise.  
Over time, when your project grows you might realize the value of component class names.

**Reply by ltbits** [reply]
Thanks for the reply @saadeghi. My point wasn't about my personal preference, but rather about idea of creating a plugin for a library that has certain stated "values" and going against those in the plugin. 

Is there any plugin or component framework coming from Tailwind core team that uses "semantic" naming? I'm not aware of one and as I pointed out their official component library uses nothing but utility class names.

**Reply by saadeghi** [reply]
@ltbits 

> Is there any plugin or component framework coming from Tailwind core team that uses "semantic" naming? I'm not aware of one and as I pointed out their official component library uses nothing but utility class names.

- [tailwindcss-forms](https://github.com/tailwindlabs/tailwindcss-forms) using semantic class names like `form-input`, `form-checkbox`. very similar to how daisyUI uses `input` or `checkbox` class names. Example: https://play.tailwindcss.com/tLjcHtSv4K
- [tailwindcss-typography](https://github.com/tailwindlabs/tailwindcss-typography) using class names like `prose` to apply styles to different elements instead of using utility classes. Example: https://play.tailwindcss.com/HxS5YibA8B
- [Headless UI](https://headlessui.com/) is using semantic names like `Checkbox`, `Button`, etc – not as class name but to name JS components – which is "semantic naming" not "utility name". Example: https://headlessui.com/react/button
- [Catalyst UI Kit](https://catalyst.tailwindui.com/docs) is using semantic names like `Checkbox`, `Button`, etc – not as class name but to name JS components – which is "semantic naming". Example: https://catalyst.tailwindui.com/docs/button

These are all products from Tailwind Team, using semantic names, either as CSS class names or as JS exports, those names are semantic names, not utility names.  

If you want to organize styles, you will end up using semantic names anyway, either by putting CSS rules under one `.checkbox` class, or by putting class names under one `<Checkbox>` JS component. That `checkbox` is a semantic name to make something look like a "checkbox" regardless of the individual CSS properties.

Even if there was no product directly from Tailwind CSS team using semantic names, you may want to think why this many people is using daisyUI, Radix, MUI, Shadcn, NuxtUI and all the other libraries that are using semantic names.  
Because maintaining huge files full of class names is not possible.  
At some point you have to put utility classes into something and put a semantic name on it. Either a `.checkbox` CSS class name or a `<Checkbox>` JS component.

**Reply by andyl** [reply]
Or a `<Checkbox>` Phoenix component...

**Reply by ltbits** [reply]
> [tailwindcss-forms](https://github.com/tailwindlabs/tailwindcss-forms) using semantic class names...
> [tailwindcss-typography](https://github.com/tailwindlabs/tailwindcss-typography) using class names like prose

Those two have very defined exceptional use cases for having less than dozen non-utility class names, as described quite well in their README. Bringing those two as examples of the Tailwind core team being OK with semantic class naming is misleading.

> why this many people is using daisyUI, Radix, MUI, Shadcn, NuxtUI 

There are "this many people" who don't use any of those. But that type of argument won't lead anywhere.

> At some point you have to put utility classes into something and put a semantic name on it.

I disagree. Maintaining huge project without semantic class names proved to be very beneficial and that is why Tailwind got so popular.

But, there might be people who prefer semantic class names; that's fine. My point isn't about one team liking utility class names only, another team liking semantic class names, or a combination of both. My point is that DaisyUI and any other plugin/library that is heavily based on semantic class naming is in conflict with the values of Tailwind and does a disservice to the Tailwind community.

If DaisyUI was advertised as a standalone thing, not a tailwind plugin (even if it uses TW internally), then it would have been a quite different story and none of my "complaints" above would exist.

**Reply by saadeghi** [reply]
> My point is that DaisyUI and any other plugin/library that is heavily based on semantic class naming is in conflict with the values of Tailwind and does a disservice to the Tailwind community.

Wait… are you saying that you are the one in charge to decide what people should use? 😅

**Reply by ltbits** [reply]
Now, where did I say such a thing? I merely tried to point out the value conflict between the two. I'm aware that pointing this out won't be welcomed in this project.

**Reply by saadeghi** [reply]
Who is deciding what is the "value"?  

To me, value is fixing a real problem.  
I see a problem with development experience of using utility-only, I tried to fix it for myself.  
Turns out many many people are having the same issue with utility-only as well.  

<img width="632" alt="Screenshot 2025-04-04 at 4 10 45 PM" src="https://github.com/user-attachments/assets/bf8c14b6-43b9-48e1-a456-55ff0dacab44" />
<img width="628" alt="Screenshot 2025-04-04 at 4 10 34 PM" src="https://github.com/user-attachments/assets/7b492026-922f-457b-878d-7e1f925bea5f" />
<img width="615" alt="Screenshot 2025-04-04 at 4 10 25 PM" src="https://github.com/user-attachments/assets/bba7d733-6e5e-4d9e-b226-407cd4d8115a" />
<img width="614" alt="Screenshot 2025-04-04 at 4 10 16 PM" src="https://github.com/user-attachments/assets/fea3f8d5-e851-44d6-a2ec-b64e961daa60" />

Not only daisyUI is in conflict to Tailwind CSS community, it actually improves the experience for people who want to use Tailwind CSS.  
To use utility-only, you need a pretty good understanding of CSS, you need to spend time and effort to work with every CSS property (= utility class) at low level.  
Component libraries with semantic approach that are built on top Tailwind, improves this experience by providing a higher level of abstraction.

If you think this is not a value, and instead the value is conservatively using utility-classes-only because of an ideology, then it seems like you're deciding what counts as value and you're not okay what people are choosing to use.

**Reply by ltbits** [reply]
I think you completely misunderstood what I meant by "value" in above statements. DaisyUI brings tremendous value to its users.
